### PR TITLE
[FIX] Shorten NLP CI tests

### PR DESCRIPTION
**Pull request by yzhang123**
* add max_steps argument to Squad in addition to num_epochs
* reduce CI test time of 
    * BERT pretrainings
    * Squad tests
    * chatbot example

### CODE REVIEW

File: examples/start_here/chatbot_example.py [hunk 1]
```diff
@@ -65,12 +65,14 @@ def outputs2words(tensors, vocab):
     tensors=[loss, src, outputs_inf, tgt], print_func=lambda x: outputs2words(x, dl.voc.index2word),
 )
 
-num_epochs = 1
-logging.info(f"Training only for {num_epochs}. Train longer (~10-20) for convergence.")
+# num_epochs = 1
+max_steps = 50
+# logging.info(f"Training only for {num_epochs}. Train longer (~10-20) for convergence.")
+logging.info(f"Training only for {max_steps} steps. Train longer (~10-20) for convergence.")
```
Comment (blisc) [inline examples/start_here/chatbot_example.py]: Can you change `Train longer (~10-20) for convergence` to `Train longer for convergence. Recommended 10-20 epochs`? Or something similar so that it is clear that it means 10~20 epochs not steps?

File: Jenkinsfile [hunk 2]
```diff
@@ -183,8 +183,8 @@ pipeline {
...
-            sh 'cd examples/nlp/question_answering && FSCORE=$(cat outputs/squadv1_roberta/log_globalrank-0_localrank-0.txt |  grep "f1" |tail -n 1 |egrep -o "[0-9.]+"|tail -n 1 ) && echo $FSCORE && if [ $(echo "$FSCORE > 50.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
+            sh 'cd examples/nlp/question_answering && CUDA_VISIBLE_DEVICES=1 python question_answering_squad.py --no_data_cache --amp_opt_level O1 --train_file /home/TestData/nlp/squad_mini/v1.1/train-v1.1.json --dev_file /home/TestData/nlp/squad_mini/v1.1/dev-v1.1.json --work_dir outputs/squadv1_roberta --batch_size 2 --save_step_freq 200 --max_steps 10 --loss_step_freq 2  --lr_policy WarmupAnnealing  --lr 1e-5 --do_lower_case  --model_type roberta --pretrained_model_name roberta-base'
+            sh 'cd examples/nlp/question_answering && FSCORE=$(cat outputs/squadv1_roberta/log_globalrank-0_localrank-0.txt |  grep "f1" |tail -n 1 |egrep -o "[0-9.]+"|tail -n 1 ) && echo $FSCORE && if [ $(echo "$FSCORE > 3.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
```
Comment (blisc) [inline Jenkinsfile]: This new F1 score (assuming that is what you're measuring) is really low

File: Jenkinsfile [hunk 3]
```diff
@@ -150,15 +150,15 @@ pipeline {
...
-            sh 'cd examples/nlp/question_answering && FSCORE=$(cat outputs/squadv1/log_globalrank-0_localrank-0.txt |  grep "f1" |tail -n 1 |egrep -o "[0-9.]+"|tail -n 1 ) && echo $FSCORE && if [ $(echo "$FSCORE > 50.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
+            sh 'cd examples/nlp/question_answering && CUDA_VISIBLE_DEVICES=0 python question_answering_squad.py --no_data_cache --amp_opt_level O1 --train_file /home/TestData/nlp/squad_mini/v1.1/train-v1.1.json --dev_file /home/TestData/nlp/squad_mini/v1.1/dev-v1.1.json --work_dir outputs/squadv1 --batch_size 8 --save_step_freq 200 --max_steps 10 --loss_step_freq 2 --lr_policy WarmupAnnealing  --lr 1e-5 --do_lower_case'
+            sh 'cd examples/nlp/question_answering && FSCORE=$(cat outputs/squadv1/log_globalrank-0_localrank-0.txt |  grep "f1" |tail -n 1 |egrep -o "[0-9.]+"|tail -n 1 ) && echo $FSCORE && if [ $(echo "$FSCORE > 4.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
```
Comment (blisc) [inline Jenkinsfile]: This new F1 score (assuming that is what you're measuring) is really low

File: Jenkinsfile [hunk 4]
```diff
@@ -150,15 +150,15 @@ pipeline {
...
         stage('BERT Squad v1.1') {
           steps {
-            sh 'cd examples/nlp/question_answering && CUDA_VISIBLE_DEVICES=0 python question_answering_squad.py --no_data_cache --amp_opt_level O1 --train_file /home/TestData/nlp/squad_mini/v1.1/train-v1.1.json --eval_file /home/TestData/nlp/squad_mini/v1.1/dev-v1.1.json --work_dir outputs/squadv1 --batch_size 8 --save_step_freq 300 --num_epochs 3 --lr_policy WarmupAnnealing  --lr 3e-5 --do_lower_case'
-            sh 'cd examples/nlp/question_answering && FSCORE=$(cat outputs/squadv1/log_globalrank-0_localrank-0.txt |  grep "f1" |tail -n 1 |egrep -o "[0-9.]+"|tail -n 1 ) && echo $FSCORE && if [ $(echo "$FSCORE > 50.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
+            sh 'cd examples/nlp/question_answering && CUDA_VISIBLE_DEVICES=0 python question_answering_squad.py --no_data_cache --amp_opt_level O1 --train_file /home/TestData/nlp/squad_mini/v1.1/train-v1.1.json --dev_file /home/TestData/nlp/squad_mini/v1.1/dev-v1.1.json --work_dir outputs/squadv1 --batch_size 8 --save_step_freq 200 --max_steps 10 --loss_step_freq 2 --lr_policy WarmupAnnealing  --lr 1e-5 --do_lower_case'
```
Comment (blisc) [inline Jenkinsfile]: Did you rename eval_file to dev_file??

File: Jenkinsfile [hunk 5]
```diff
@@ -65,15 +65,15 @@ pipeline {
...
-            sh 'cd examples/nlp/language_modeling && LOSS=$(cat outputs/bert_lm/wiki_book/log_globalrank-0_localrank-0.txt |  grep "Loss" |tail -n 1| awk \'{print \$7}\' | egrep -o "[0-9.]+" ) && echo $LOSS && if [ $(echo "$LOSS < 15.0" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
+            sh 'cd examples/nlp/language_modeling && CUDA_VISIBLE_DEVICES=1 python bert_pretraining.py --amp_opt_level O1 --data_dir /home/TestData/nlp/wiki_book_mini  --work_dir outputs/bert_lm/wiki_book --batch_size 8 --config_file /home/TestData/nlp/bert_configs/uncased_L-12_H-768_A-12.json  --save_step_freq 100 --max_steps 5 --print_step_freq 1  --num_gpus 1  --batches_per_step 1 --lr_policy SquareRootAnnealing --beta2 0.999 --beta1 0.9  --lr_warmup_proportion 0.01 --optimizer adam_w  --weight_decay 0.01  --lr 0.875e-4 --preprocessed_data '
+            sh 'cd examples/nlp/language_modeling && LOSS=$(cat outputs/bert_lm/wiki_book/log_globalrank-0_localrank-0.txt |  grep "Loss" |head -n 1| awk \'{print \$7}\' | egrep -o "[0-9.]+" ) && echo $LOSS && if [ $(echo "$LOSS < 11.5" | bc -l) -eq 1 ]; then echo "SUCCESS" && exit 0; else echo "FAILURE" && exit 1; fi'
```
Comment (blisc) [inline Jenkinsfile]: should this be tail or head?

File: examples/start_here/chatbot_example.py (hunk 1 above)
Comment (yzhang123) [reply examples/start_here/chatbot_example.py]: ok

File: Jenkinsfile (hunk 4 above)
Comment (blisc) [reply Jenkinsfile]: Ok you changed it back, so I assume that you didn't

File: Jenkinsfile (hunk 2 above)
Comment (yzhang123) [reply Jenkinsfile]: yes, it just a dummy test to check if the F1 is not 0.

File: Jenkinsfile (hunk 3 above)
Comment (yzhang123) [reply Jenkinsfile]: yes, its just a dummy test

File: Jenkinsfile (hunk 4 above)
Comment (yzhang123) [reply Jenkinsfile]: fixed this typo

File: Jenkinsfile (hunk 5 above)
Comment (yzhang123) [reply Jenkinsfile]: I'm using head now since its more deterministic

File: Jenkinsfile (hunk 2 above)
Comment (blisc) [reply Jenkinsfile]: If it is a dummy test, then what is the value of the test? I would argue that it is more important to have a stricter CI than it is to save CI test time.
@tkornuta-nvidia, do you have an opinion here?

File: Jenkinsfile (hunk 5 above)
Comment (blisc) [reply Jenkinsfile]: Doesn't head return the first element in the list? Ie you would be comparing the loss at step 0 rather than the last step. Tail seems like it makes the most sense

Comment (blisc) [reply Jenkinsfile]: Maybe we can go to the other extreme and just limit our dataset to our batch size. IE do gradient descent inside of SGD

File: Jenkinsfile (hunk 2 above)
Comment (blisc) [reply Jenkinsfile]: I think that testing that loss is non-zero is a L0 or unittest

Comment (tkornuta-nvidia) [reply Jenkinsfile]: I agree that checking if F1 is not 0 has a low "information value" - at that point I would change really like to change that to forward-backward step. Then it will be L0.

Finally, I really think the whole "testing logic" (e.g. checking a condition) should be implemented **outside** of Jenkins configuration file. This looks like a hack, whereas python unittest provides tools devoted exactly for that purpose. This approach will make things harder to manage/reorganize/use in a long run...

File: Jenkinsfile (hunk 3 above)
Comment (tkornuta-nvidia) [reply Jenkinsfile]: ech, the same here, seems the pattern already "propagated"...

File: Jenkinsfile (hunk 5 above)
Comment (tkornuta-nvidia) [reply Jenkinsfile]: the same :(

File: Jenkinsfile (hunk 2 above)
Comment (yzhang123) [reply Jenkinsfile]: hey @tkornuta-nvidia @blisc, I have updated the criteria for the ci tests regarding your feedback, the new F1 thresh will be around ~10, which is higher than in step 0 which is ~4-7.
This diff is outdated

**Comment by blisc** [comment]
> 1. Move the "testing logic" from Jenkins to python
Can you elaborate on the benefits of this? The python file would be doing exactly what the shell script is doing. Opening a file, reading it, and comparing loss.
Regarding your comment on python unittests, we are not using unittests for CI, nor do I think we should. I find it much more beneficial that CI tests are actually running our examples so we can catch typos and other issues that might be present inside the example python files themselves.

**Comment by yzhang123** [comment]
> 1. Move the "testing logic" from Jenkins to python
> 2. Those tests look like "Integration tests" - you are not really interested whether the model converges or not, just checking whether the modules are properly connected and work when combined into a graph. For me this is a perfect case for a single forward (eventually backward) step.

These are system tests, since I do check if the model improves

**Comment by okuchaiev** [comment]
I'm not sure if we should "pollute" examples with testing logic

**Comment by tkornuta-nvidia** [comment]
For sure we shouldn't "pollute" our examples with code related to testing.

My point is: we should stop using "examples" as "tests" - at least when it comes to unit, integration and system level. They serve different purpose.

I see the origin of that problem - generally "the code is almost the same, maybe aside of assert". I think once we abstract generation of graphs into configuration files, we will be able to "automate" plenty of system tests by telling: "load config x, train for y, check if loss went down".
